File: code/appendix/torch_tansacnet/lsunAnalysis2dNetwork.py
import torch
import torch.nn as nn
from .lsunBlockDct2dLayer import LsunBlockDct2dLayer 
from .lsunInitialRotation2dLayer import LsunInitialRotation2dLayer 
from .lsunAtomExtension2dLayer import LsunAtomExtension2dLayer
from .lsunIntermediateRotation2dLayer import LsunIntermediateRotation2dLayer
from .lsunChannelSeparation2dLayer import LsunChannelSeparation2dLayer
from .lsunLayerExceptions import InvalidOverlappingFactor, InvalidNoDcLeakage, InvalidNumberOfLevels, InvalidStride, InvalidInputSize
from .lsunUtility import Direction

class LsunAnalysis2dNetwork(nn.Module):
    """
    LSUNANALYSIS2DNETWORK
    
    Requirements: Python 3.10-12.x, PyTorch 2.3/4.x
    
    Copyright (c) 2024, Shogo MURAMATSU
    
    All rights reserved.
    
    Contact address: Shogo MURAMATSU,
        Faculty of Engineering, Niigata University,
        8050 2-no-cho Ikarashi, Nishi-ku,
        Niigata, 950-2181, JAPAN
    
        https://www.eng.niigata-u.ac.jp/~msiplab/
    """
    def __init__(self,
        input_size=[2, 2], 
        number_of_components=1,
        stride=[2, 2],
        overlapping_factor=[1,1],
        no_dc_leakage=True,
        number_of_levels=0,
        prefix='',
        dtype=torch.get_default_dtype(),
        device=torch.get_default_device()
        ):
        super(LsunAnalysis2dNetwork, self).__init__()

        # Check and set parameters
        self.dtype = dtype
        self.device = device
     
        # Stride
        nDecs = stride[Direction.VERTICAL]*stride[Direction.HORIZONTAL]
        if nDecs%2!=0:
            raise InvalidStride(
            '%d x %d : Currently, even product of strides is only supported.'\
            % (stride[Direction.VERTICAL], stride[Direction.HORIZONTAL]))
        self.stride = stride

        # Overlapping factor
        if (overlapping_factor[Direction.VERTICAL]-1)%2 or (overlapping_factor[Direction.HORIZONTAL]-1)%2: 
             raise InvalidOverlappingFactor(
             '%d + %d : Currently, odd overlapping factors are only supported.'\
             % (overlapping_factor[Direction.VERTICAL], overlapping_factor[Direction.HORIZONTAL]))
        self.overlapping_factor = overlapping_factor

        # No DC leakage
        if not isinstance(no_dc_leakage, bool):
            raise InvalidNoDcLeakage("no_dc_leakage must be a boolean value")
        self.no_dc_leakage = no_dc_leakage

        # # of levels
        if not isinstance(number_of_levels, int):
            raise InvalidNumberOfLevels(
            '%f : The number of levels must be integer.'\
            % number_of_levels)   
        if number_of_levels < 0:
            raise InvalidNumberOfLevels(
            '%d : The number of levels must be greater than or equal to 0.'\
            % number_of_levels)
        self.number_of_levels = number_of_levels

        # # of blocks
        if input_size[Direction.VERTICAL]%stride[Direction.VERTICAL] != 0 or input_size[Direction.HORIZONTAL]%stride[Direction.HORIZONTAL] != 0:
            raise InvalidInputSize(
            '%d x %d : Currently, multiples of strides is only supported.'\
            % (input_size[Direction.VERTICAL], input_size[Direction.HORIZONTAL]))
        if input_size[Direction.VERTICAL] < 1 or input_size[Direction.HORIZONTAL] < 1:
            raise InvalidInputSize(
            '%d x %d : Positive integers are only supported.'\
            % (input_size[Direction.VERTICAL], input_size[Direction.HORIZONTAL]))
        nrows = input_size[Direction.VERTICAL]//stride[Direction.VERTICAL]
        ncols = input_size[Direction.HORIZONTAL]//stride[Direction.HORIZONTAL]
        self.input_size = input_size

        # Instantiation of layers
        if self.number_of_levels == 0:
            nlevels = 1
        else:
            nlevels = self.number_of_levels

        stages = [ nn.Sequential() for iStage in range(nlevels) ]
        for iStage in range(len(stages)):
            iLevel = iStage + 1
            strLv = 'Lv%0d_'%iLevel

            # Initial blocks
            stages[iStage].add_module(prefix+strLv+'E0',LsunBlockDct2dLayer(
                stride=self.stride
                ))

            stages[iStage].add_module(prefix+strLv+'V0',LsunInitialRotation2dLayer(
                stride=self.stride,
                number_of_blocks=[nrows,ncols], 
                mus=1,               
                no_dc_leakage=self.no_dc_leakage,
                dtype=self.dtype,device=self.device))

            # Horizontal extension
            for iOrderH in range(2,overlapping_factor[Direction.HORIZONTAL],2):
                stages[iStage].add_module(prefix+strLv+'Qh%drd'%(iOrderH-1),LsunAtomExtension2dLayer(
                    stride=self.stride,
                    direction='Right',
                    target_channels='Difference'))
                stages[iStage].add_module(prefix+strLv+'Vh%d'%(iOrderH-1),LsunIntermediateRotation2dLayer(
                    stride=self.stride,
                    number_of_blocks=[nrows,ncols],                                         
                    mode='Analysis',
                    mus=-1,
                    dtype=self.dtype,device=self.device))
                stages[iStage].add_module(prefix+strLv+'Qh%dls'%iOrderH,LsunAtomExtension2dLayer(
                    stride=self.stride,
                    direction='Left',
                    target_channels='Sum'))
                stages[iStage].add_module(prefix+strLv+'Vh%d'%iOrderH,LsunIntermediateRotation2dLayer(
                    stride=self.stride,
                    number_of_blocks=[nrows,ncols],                     
                    mode='Analysis',
                    mus=-1,
                    dtype=self.dtype,device=self.device))
                
            # Vertical extension
            for iOrderV in range(2,overlapping_factor[Direction.VERTICAL],2):            
                stages[iStage].add_module(prefix+strLv+'Qv%ddd'%(iOrderV-1),LsunAtomExtension2dLayer(
                    stride=self.stride,
                    direction='Down',
                    target_channels='Difference'))                
                stages[iStage].add_module(prefix+strLv+'Vv%d'%(iOrderV-1),LsunIntermediateRotation2dLayer(
                    stride=self.stride,
                    number_of_blocks=[nrows,ncols],                                             
                    mode='Analysis',
                    mus=-1,
                    dtype=self.dtype,device=self.device))
                stages[iStage].add_module(prefix+strLv+'Qv%dus'%iOrderV,LsunAtomExtension2dLayer(
                    stride=self.stride,
                    direction='Up',
                    target_channels='Sum'))     
                stages[iStage].add_module(prefix+strLv+'Vv%d'%iOrderV,LsunIntermediateRotation2dLayer(
                    stride=self.stride,
                    number_of_blocks=[nrows,ncols],                         
                    mode='Analysis',
                    mus=-1,
                    dtype=self.dtype,device=self.device))
 
            # Channel Separation for intermediate stages
            if self.number_of_levels > 0:
                stages[iStage].add_module(strLv+'Sp',LsunChannelSeparation2dLayer())

            # Update size
            nrows = nrows//stride[Direction.VERTICAL]
            ncols = ncols//stride[Direction.HORIZONTAL]    

        # Stack modules as a list
        self.layers = nn.ModuleList(stages)
        
    def forward(self,x):
        
        if self.device.type == 'cuda':
            return self.forward_cuda(x)
        else:
            return self.forward_cpu(x)
           
    def forward_cuda(self,x):   
        # TODO: #10 Stream processing

        if self.number_of_levels == 0: # Flat structure
            m = self.layers[0]
            x = m.forward(x)
            return x
        else: # Tree structure
            stride = self.stride
            nSamples = x.size(0)
            nComponents = x.size(1)
            nrows = x.size(2)//stride[Direction.VERTICAL]
            ncols = x.size(3)//stride[Direction.HORIZONTAL]
            y = []
            iLevel = 1       
            for m in self.layers:
                yac, ydc = m.forward(x)
                y.insert(0,yac)
                if iLevel < self.number_of_levels:
                    x = ydc.view(nSamples,nComponents,nrows,ncols)
                    nrows = nrows//stride[Direction.VERTICAL]
                    ncols = ncols//stride[Direction.HORIZONTAL]   
                    iLevel += 1
                else:
                    y.insert(0,ydc.unsqueeze(3))
            return tuple(y)        
        
    def forward_cpu(self,x):
        
        if self.number_of_levels == 0: # Flat structure
            m = self.layers[0]
            x = m.forward(x)
            return x
        else: # Tree structure
            stride = self.stride
            nSamples = x.size(0)
            nComponents = x.size(1)
            nrows = x.size(2)//stride[Direction.VERTICAL]
            ncols = x.size(3)//stride[Direction.HORIZONTAL]
            y = []
            iLevel = 1       
            for m in self.layers:
                yac, ydc = m.forward(x)
                y.insert(0,yac)
                if iLevel < self.number_of_levels:
                    x = ydc.view(nSamples,nComponents,nrows,ncols)
                    nrows = nrows//stride[Direction.VERTICAL]
                    ncols = ncols//stride[Direction.HORIZONTAL]   
                    iLevel += 1
                else:
                    y.insert(0,ydc.unsqueeze(3))
            return tuple(y)     
        
    @property
    def T(self):
        from .lsunSynthesis2dNetwork import LsunSynthesis2dNetwork
        import re

        # Create synthesizer as the adjoint of SELF
        synthesizer = LsunSynthesis2dNetwork(
            input_size=self.input_size,
            stride=self.stride,
            overlapping_factor=self.overlapping_factor,
            no_dc_leakage=self.no_dc_leakage,
            number_of_levels=self.number_of_levels,
            dtype=self.dtype,
            device=self.device
            )
        
        if self.number_of_levels == 0:
            nlevels = 1
        else:
            nlevels = self.number_of_levels

        # Copy state dictionary
        ana_state_dict = self.state_dict()
        syn_state_dict = synthesizer.state_dict()
        for key in syn_state_dict.keys():
            istage_ana = int(re.sub('^layers\.|\.Lv\d_.+$','',key))
            istage_syn = (nlevels-1)-istage_ana
            angs = ana_state_dict[key.replace('layers.%d'%istage_ana,'layers.%d'%istage_syn).replace('~','').replace('T.orthonormalTransforms','.orthonormalTransforms') ] 
            syn_state_dict[key] = angs
        
        # Load state dictionary
        synthesizer.load_state_dict(syn_state_dict)

        # Return adjoint
        return synthesizer #.to(dtype=self.dtype,device=self.device)

    def to(self, device=None, dtype=None,*args, **kwargs):
        if device is not None:
            self.device = device
        if dtype is not None:
            self.dtype = dtype
        self = super(LsunAnalysis2dNetwork, self).to(device=self.device,dtype=self.dtype,*args, **kwargs)
        for s in self.layers:
            for m in s:
                m.to(device=self.device,dtype=self.dtype,*args, **kwargs)
        return self

File: code/+tansacnet/+lsun/lsunAnalysis2dNetwork.m
classdef lsunAnalysis2dNetwork < dlnetwork
    %LSUNANALYSIS2dNETWORK
    %
    % Requirements: MATLAB R2024a
    %
    % Copyright (c) 2025, Shogo MURAMATSU, Yasas GODAGE
    %
    % All rights reserved.
    %
    % Contact address: Shogo MURAMATSU
    %                Faculty of Engineering, Niigata University,
    %                8050 2-no-cho Ikarashi, Nishi-ku,
    %                Niigata, 950-2181, JAPAN
    %


    properties
        % fcn_createlsunlgraph2d 参照
        TensorSize
        Stride
        OverlappingFactor
        NumberOfVanishingMoments
        NumberOfLevels
        NoDcLeakage
        Prefix
        Device
        DType
    end

    properties (Access = private)
        % Private members
        PrivateInputSize
        PrivateNumberOfComponents
    end

    methods
        function obj = lsunAnalysis2dNetwork(varargin)
            %UNTITLED このクラスのインスタンスを作成
            %   詳細説明をここに記述
            
            % fcn_createlsunlgraph2d のプロパティ設定を参照
            p = inputParser;
            addParameter(p,'InputSize',[32 32]);
            addParameter(p,'NumberOfComponents',1);
            addParameter(p,'NumberOfVanishingMoments',[1 1]);
            addParameter(p,'Stride',[2 2]);
            addParameter(p,'OverlappingFactor',[1 1]);
            addParameter(p,'NumberOfLevels',1);
            addParameter(p,'NoDcLeakage',false);
            addParameter(p,'Prefix','');
            addParameter(p,'Device','cuda');
            addParameter(p,'DType','double');
            parse(p,varargin{:});

            % Parameters
            obj.PrivateInputSize = p.Results.InputSize;
            obj.PrivateNumberOfComponents = p.Results.NumberOfComponents;
            obj.TensorSize = [obj.PrivateInputSize obj.PrivateNumberOfComponents];  
            obj.Stride = p.Results.Stride;
            obj.OverlappingFactor = p.Results.OverlappingFactor;
            obj.NumberOfVanishingMoments = p.Results.NumberOfVanishingMoments;
            obj.NumberOfLevels = p.Results.NumberOfLevels;
            obj.NoDcLeakage = p.Results.NoDcLeakage;
            obj.Prefix = p.Results.Prefix;
            obj.Device = p.Results.Device;
            obj.DType = p.Results.DType;
        end

        function analyzerFactory = getAnalyzerFactory(obj)
            % ???      
            analyzerFactory = [];
        end

        function adj = transpose(obj)
            %METHOD1 このメソッドの概要をここに記述
            %   詳細説明をここに記述

            % fcn_cpparamsana2syn を参照

            % import tansacnet.lsun.*
            % %
            % expanalyzer = '^Lv\d+_Cmp\d+_V(\w\d|0)+$';
            % nLayers = height(analysisdlnet.Layers);
            % for iLayer = 1:nLayers
            %     alayer = analysisdlnet.Layers(iLayer);
            %     alayerName = alayer.Name;
            %     if ~isempty(regexp(alayerName,expanalyzer,'once'))
            %         slayer = synthesislgraph.Layers({synthesislgraph.Layers.Name} == alayerName + "~");
            %         alayer.Angles = slayer.Angles;
            %         alayer.Mus = slayer.Mus;
            %         if isa(alayer,'tansacnet.lsun.lsunInitialRotation2dLayer')
            %             alayer.NoDcLeakage = slayer.NoDcLeakage;
            %         end
            %         analysisdlnet = analysisdlnet.replaceLayer(alayerName,alayer);
            %         disp("Copy angles from " + slayer.Name + " to " + alayerName)
            %     end
            % end
        end
    end

    methods (Hidden=false)
        function dlnet = dlnetwork(obj)
            
            % lsundlnet = [];
            % if isempty(lsundlnet)
            %     lsundlnet = dlnetwork;
            % end
            lsundlnet = dlnetwork;

            nComponents = obj.PrivateNumberOfComponents;
            inputSize = [obj.PrivateInputSize nComponents];
            stride = obj.Stride;
            ovlpFactor = obj.OverlappingFactor;
            nLevels = obj.NumberOfLevels;
            %
            device = obj.Device;
            dtype = obj.DType;
            %
            noDcLeakage = obj.NumberOfVanishingMoments;

            if isscalar(noDcLeakage)
                noDcLeakage = [1 1]*noDcLeakage; 
            end

            prefix = obj.Prefix;
            %isapndinout = obj.AppendInOutLayers;

            nDecs = prod(stride);
            nChannels = [ceil(nDecs/2) floor(nDecs/2)];
            
            if nChannels(1) ~= nChannels(2)
                throw(MException('LsunLayer:InvalidStride',...
                    '[%d %d] : The product of stride should be even.',...
                    stride(1),stride(2)))
            end
            if ~all(mod(ovlpFactor,2))
                throw(MException('LsunLayer:InvalidOverlappingFactor',...
                    '[%d %d] : Currently, odd overlapping factors are only supported.',...
                    ovlpFactor(1),ovlpFactor(2)))
            end

            nBlocks = inputSize(1:2)./stride;
            if any(mod(nBlocks,1))
                error('[%d %d] : Input size should be multiple of stride.')
            end

            blockDctLayers = cell(nLevels);
            analysisLayers = cell(nLevels,nComponents);
            %blockIdctLayers = cell(nLevels);
            %synthesisLayers = cell(nLevels,nComponents);
            import tansacnet.lsun.*
            for iLv = 1:nLevels
                strLv = sprintf('Lv%0d_',iLv);
                
                % Initial blocks
                blockDctLayers{iLv} = lsunBlockDct2dLayer('Name',[prefix strLv 'E0'],...
                    'Stride',stride,...
                    'NumberOfComponents',nComponents);
                for iCmp = 1:nComponents
                    strCmp = sprintf('Cmp%0d_',iCmp);
                    analysisLayers{iLv,iCmp} = [ analysisLayers{iLv,iCmp}
                        lsunInitialRotation2dLayer('Name',[prefix strLv strCmp 'V0'],...
                        'NumberOfBlocks',nBlocks,'Stride',stride,...
                        'NoDcLeakage',noDcLeakage(1),...
                         'Device',device,...
                         'DType',dtype)
                        ];
                end

                
                % Final blocks
                % blockIdctLayers{iLv} = lsunBlockIdct2dLayer('Name',[prefix strLv 'E0~'],...
                %     'Stride',stride,...
                %     'NumberOfComponents',nComponents);
                % for iCmp = 1:nComponents
                %     strCmp = sprintf('Cmp%0d_',iCmp);
                %     synthesisLayers{iLv,iCmp} = [ synthesisLayers{iLv,iCmp}
                %         lsunFinalRotation2dLayer('Name',[prefix strLv strCmp 'V0~'],...
                %         'NumberOfBlocks',nBlocks,'Stride',stride,...
                %         'NoDcLeakage',noDcLeakage(2),...
                %             'Device',device,...
                %             'DType',dtype)
                %         ];
                % end
                
                % NOTE on the modification of parameter Mus 
                %
                % Modified default parameter settings so that the default setting becomes a block DCT.
                % Note that in addition to the change in default design values, design parameters are no longer compatible with previous releases.
                % Redesign is strongly recommended!
                %
                % Reference:
                % S. Muramatsu, T. Kobayashi, M. Hiki and H. Kikuchi, "Boundary Operation of 2-D Nonseparable Linear-Phase Paraunitary Filter Banks," in IEEE Transactions on Image Processing, vol. 21, no. 4, pp. 2314-2318, April 2012, doi: 10.1109/TIP.2011.2181527.
                %
                % Date: 6 Setp. 2024
                %
                for iCmp = 1:nComponents
                    strCmp = sprintf('Cmp%0d_',iCmp);
                    % Atom extension in horizontal
                    for iOrderH = 2:2:ovlpFactor(2)-1
                        analysisLayers{iLv,iCmp} = [ analysisLayers{iLv,iCmp}
                            lsunAtomExtension2dLayer('Name',[prefix strLv strCmp 'Qh' num2str(iOrderH-1) 'rd'],...
                            'Stride',stride,'Direction','Right','TargetChannels','Difference')
                            lsunIntermediateRotation2dLayer('Name',[prefix strLv strCmp 'Vh' num2str(iOrderH-1)],...
                            'Stride',stride,'NumberOfBlocks',nBlocks,'Mode','Analysis','Mus',-1,...
                                'Device',device,'DType',dtype)
                            lsunAtomExtension2dLayer('Name',[prefix strLv strCmp 'Qh' num2str(iOrderH) 'ls'],...
                            'Stride',stride,'Direction','Left','TargetChannels','Sum')
                            lsunIntermediateRotation2dLayer('Name',[prefix strLv strCmp 'Vh' num2str(iOrderH) ],...
                            'Stride',stride,'NumberOfBlocks',nBlocks,'Mode','Analysis','Mus',-1,...
                                'Device',device,...
                                'DType',dtype) % Revised default Mus to -1 on 6 Sept. 2024
                            ];
                        % synthesisLayers{iLv,iCmp} = [ synthesisLayers{iLv,iCmp}
                        %     lsunAtomExtension2dLayer('Name',[prefix strLv strCmp 'Qh' num2str(iOrderH-1) 'rd~'],...
                        %     'Stride',stride,'Direction','Left','TargetChannels','Difference')
                        %     lsunIntermediateRotation2dLayer('Name',[prefix strLv strCmp 'Vh' num2str(iOrderH-1) '~'],...
                        %     'Stride',stride,'NumberOfBlocks',nBlocks,'Mode','Synthesis','Mus',-1,...
                        %         'Device',device,...
                        %         'DType',dtype)
                        %     lsunAtomExtension2dLayer('Name',[prefix strLv strCmp 'Qh' num2str(iOrderH) 'ls~'],...
                        %     'Stride',stride,'Direction','Right','TargetChannels','Sum')
                        %     lsunIntermediateRotation2dLayer('Name',[prefix strLv strCmp 'Vh' num2str(iOrderH) '~'],...
                        %     'Stride',stride,'NumberOfBlocks',nBlocks,'Mode','Synthesis','Mus',-1,...
                        %         'Device',device,...
                        %         'DType',dtype) % Revised default Mus to -1 on 6 Sept. 2024
                        %     ];
                    end
                    % Atom extension in vertical
                    for iOrderV = 2:2:ovlpFactor(1)-1
                        analysisLayers{iLv,iCmp} = [ analysisLayers{iLv,iCmp}
                            lsunAtomExtension2dLayer('Name',[prefix strLv strCmp 'Qv' num2str(iOrderV-1) 'dd'],...
                            'Stride',stride,'Direction','Down','TargetChannels','Difference')
                            lsunIntermediateRotation2dLayer('Name',[prefix strLv strCmp 'Vv' num2str(iOrderV-1)],...
                            'Stride',stride,'NumberOfBlocks',nBlocks,'Mode','Analysis','Mus',-1,...
                                'Device',device,...
                                'DType',dtype)
                            lsunAtomExtension2dLayer('Name',[prefix strLv strCmp 'Qv' num2str(iOrderV) 'us'],...
                            'Stride',stride,'Direction','Up','TargetChannels','Sum')
                            lsunIntermediateRotation2dLayer('Name',[prefix strLv strCmp 'Vv' num2str(iOrderV)],...
                            'Stride',stride,'NumberOfBlocks',nBlocks,'Mode','Analysis','Mus',-1,...
                                'Device',device,...
                                'DType',dtype) % Revised default Mus to -1 on 6 Sept. 2024
                            ];
                        % synthesisLayers{iLv,iCmp} = [ synthesisLayers{iLv,iCmp}
                        %     lsunAtomExtension2dLayer('Name',[prefix strLv strCmp 'Qv' num2str(iOrderV-1) 'dd~'],...
                        %     'Stride',stride,'Direction','Up','TargetChannels','Difference')
                        %     lsunIntermediateRotation2dLayer('Name',[prefix strLv strCmp 'Vv' num2str(iOrderV-1) '~'],...
                        %     'Stride',stride,'NumberOfBlocks',nBlocks,'Mode','Synthesis','Mus',-1,...
                        %         'Device',device,...
                        %         'DType',dtype)
                        %     lsunAtomExtension2dLayer('Name',[prefix strLv strCmp 'Qv' num2str(iOrderV) 'us~'],...
                        %     'Stride',stride,'Direction','Down','TargetChannels','Sum')
                        %     lsunIntermediateRotation2dLayer('Name',[prefix strLv strCmp 'Vv' num2str(iOrderV) '~'],...
                        %     'Stride',stride,'NumberOfBlocks',nBlocks,'Mode','Synthesis','Mus',-1,...
                        %         'Device',device,...
                        %         'DType',dtype) % Revised default Mus to -1 on 6 Sept. 2024
                        %     ];
                    end
                    
                    % Channel separation and concatenation
                    analysisLayers{iLv,iCmp} = [ analysisLayers{iLv,iCmp}
                      lsunChannelSeparation2dLayer('Name',[prefix strLv strCmp 'Sp'])
                        ];


                    % synthesisLayers{iLv,iCmp} = [ synthesisLayers{iLv,iCmp}
                    %     lsunChannelConcatenation2dLayer('Name',[prefix strLv strCmp 'Cn'])
                    %     ];               
                end
                
            end
            
            iLv = 1;
            strLv = sprintf('Lv%0d_',iLv);

            lsundlnet = addLayers(lsundlnet,[imageInputLayer(inputSize,'Name',[prefix 'Image input'],'Normalization','none'), ...
            lsunIdentityLayer('Name',[prefix strLv 'In']), blockDctLayers{iLv}]);

            %lsundlnet = addLayers(lsundlnet,[lsunIdentityLayer('Name',[prefix strLv 'In']), blockDctLayers{iLv}]);

            for iCmp = 1:nComponents
                strCmp = sprintf('Cmp%0d_',iCmp);
                lsundlnet = addLayers(lsundlnet,analysisLayers{iLv,iCmp});
                if nComponents > 1
                    lsundlnet = connectLayers(lsundlnet,...
                        [prefix strLv 'E0/out' num2str(iCmp)], [prefix strLv strCmp 'V0']);
                else
                    lsundlnet = connectLayers(lsundlnet,...
                        [prefix strLv 'E0'], [prefix strLv strCmp 'V0']);
                end
            end

            if nComponents > 1
                lsundlnet = addLayers(lsundlnet,...
                    depthConcatenationLayer(nComponents,'Name',[prefix strLv 'AcOut']));
                lsundlnet = addLayers(lsundlnet,...
                    depthConcatenationLayer(nComponents,'Name',[prefix strLv 'DcOut']));
            else
                lsundlnet = addLayers(lsundlnet,...
                    lsunIdentityLayer('Name',[prefix strLv 'AcOut']));
                lsundlnet = addLayers(lsundlnet,...
                    lsunIdentityLayer('Name',[prefix strLv 'DcOut']));
            end

            if nComponents > 1
                for iCmp = 1:nComponents
                    strCmp = sprintf('Cmp%0d_',iCmp);
                    lsundlnet = connectLayers(lsundlnet,...
                        [prefix strLv strCmp 'Sp/ac'], [prefix strLv 'AcOut/in' num2str(iCmp)]);
                    lsundlnet = connectLayers(lsundlnet,...
                        [prefix strLv strCmp 'Sp/dc'], [prefix strLv 'DcOut/in' num2str(iCmp)]);
                end
            else
                strCmp = 'Cmp1_';
                lsundlnet = connectLayers(lsundlnet,...
                    [prefix strLv strCmp 'Sp/ac'], [prefix strLv 'AcOut/in' ]);
                lsundlnet = connectLayers(lsundlnet,...
                    [prefix strLv strCmp 'Sp/dc'], [prefix strLv 'DcOut/in' ]);
            end

            % Level n > 1
            for iLv = 2:nLevels
                strLv = sprintf('Lv%0d_',iLv);
                strLvPre = sprintf('Lv%0d_',iLv-1);
                lsundlnet = addLayers(lsundlnet,[
                    lsunIdentityLayer('Name',[prefix strLv 'In']),...
                    blockDctLayers{iLv}]);
                lsundlnet = connectLayers(lsundlnet,[prefix strLvPre 'DcOut'],[prefix strLv 'In']);
                for iCmp = 1:nComponents
                    strCmp = sprintf('Cmp%0d_',iCmp);
                    lsundlnet = addLayers(lsundlnet,analysisLayers{iLv,iCmp});
                    if nComponents > 1
                        lsundlnet = connectLayers(lsundlnet,...
                            [prefix strLv 'E0/out' num2str(iCmp)], [prefix strLv strCmp 'V0']);
                    else
                        lsundlnet = connectLayers(lsundlnet,...
                            [prefix strLv 'E0'], [prefix strLv strCmp 'V0']);
                    end
                end
                % Output
                if nComponents > 1
                    lsundlnet = addLayers(lsundlnet,...
                        depthConcatenationLayer(nComponents,'Name',[prefix strLv 'AcOut']));
                    lsundlnet = addLayers(lsundlnet,...
                        depthConcatenationLayer(nComponents,'Name',[prefix strLv 'DcOut']));
                else
                    lsundlnet = addLayers(lsundlnet,...
                        lsunIdentityLayer('Name',[prefix strLv 'AcOut']));
                    lsundlnet = addLayers(lsundlnet,...
                        lsunIdentityLayer('Name',[prefix strLv 'DcOut']));
                end
                if nComponents > 1
                    for iCmp = 1:nComponents
                        strCmp = sprintf('Cmp%0d_',iCmp);
                        lsundlnet = connectLayers(lsundlnet,...
                            [prefix strLv strCmp 'Sp/ac'], [prefix strLv 'AcOut/in' num2str(iCmp)]);
                        lsundlnet = connectLayers(lsundlnet,...
                            [prefix strLv strCmp 'Sp/dc'], [prefix strLv 'DcOut/in' num2str(iCmp)]);
                    end
                else
                    strCmp = 'Cmp1_';
                    lsundlnet = connectLayers(lsundlnet,...
                        [prefix strLv strCmp 'Sp/ac'], [prefix strLv 'AcOut/in' ]);
                    lsundlnet = connectLayers(lsundlnet,...
                        [prefix strLv strCmp 'Sp/dc'], [prefix strLv 'DcOut/in' ]);
                end
            end

            dlnet = lsundlnet;

           % ???
            %dlnet = [];
        end
    end

end
